cpp/src/io/utilities/file_io_utilities.cpp

```diff
@@ -166,28 +168,67 @@ void cufile_registered_file::register_handle()
 cufile_registered_file::~cufile_registered_file() { shim->handle_deregister(cf_handle); }
 
 cufile_input_impl::cufile_input_impl(std::string const& filepath)
-  : shim{cufile_shim::instance()}, cf_file(shim, filepath, O_RDONLY | O_DIRECT)
+  : shim{cufile_shim::instance()}, cf_file(shim, filepath, O_RDONLY | O_DIRECT), pool(16)
 {
+  pool.sleep_duration = 10;
 }
 
 std::unique_ptr<datasource::buffer> cufile_input_impl::read(size_t offset,
                                                             size_t size,
                                                             rmm::cuda_stream_view stream)
 {
   rmm::device_buffer out_data(size, stream);
-  CUDF_EXPECTS(shim->read(cf_file.handle(), out_data.data(), size, offset, 0) != -1,
-               "cuFile error reading from a file");
-
+  read(offset, size, reinterpret_cast<uint8_t*>(out_data.data()), stream);
+  // TODO: make use of the returned size to shrink out_data
   return datasource::buffer::create(std::move(out_data));
 }
 
+std::future<size_t> cufile_input_impl::read_async(size_t offset,
+                                                  size_t size,
+                                                  uint8_t* dst,
+                                                  rmm::cuda_stream_view stream)
+{
+  int device;
+  cudaGetDevice(&device);
+
+  auto read_slice = [=](void* dst, size_t size, size_t offset) -> ssize_t {
+    cudaSetDevice(device);
+    auto read_size = shim->read(cf_file.handle(), dst, size, offset, 0);
+    CUDF_EXPECTS(read_size != -1, "cuFile error reading from a file");
+    return read_size;
+  };
+
+  std::vector<std::future<ssize_t>> slice_tasks;
+  constexpr size_t four_MB = 4 * 1024 * 1024;
+  size_t n_slices          = util::div_rounding_up_safe(size, four_MB);
+  size_t slice_size        = four_MB;
+  size_t slice_offset      = 0;
+  for (size_t t = 0; t < n_slices; ++t) {
+    void* dst_slice = dst + slice_offset;
+
+    if (t == n_slices - 1) { slice_size = size % four_MB; }
+    slice_tasks.push_back(pool.submit(read_slice, dst_slice, slice_size, offset + slice_offset));
+
+    slice_offset += slice_size;
+  }
+  auto waiter = [](decltype(slice_tasks) slice_tasks) -> size_t {
+    return std::accumulate(slice_tasks.begin(), slice_tasks.end(), 0, [](auto sum, auto& task) {
+      return sum + task.get();
+    });
+  };
+  return std::async(std::launch::deferred, waiter, std::move(slice_tasks));
```

Comment:
maybe a comment on why waiter is deffered and not async

Reply:
The point of a thread pool is that we don't want to incur the overhead of thread creation in each read call. This would cause that. Significant in the case of multiple small reads.